packages/gatsby/cache-dir/app.js

```diff
@@ -34,23 +34,38 @@ apiRunnerAsync(`onClientEntry`).then(() => {
     .then(res => res.json())
     .then(services => {
       if (services.developstatusserver) {
+        let isRestarting = false
         const parentSocket = io(
           `http://${window.location.hostname}:${services.developstatusserver.port}`
         )
 
-        parentSocket.on(`develop:needs-restart`, msg => {
+        parentSocket.on(`structured-log`, msg => {
           if (
-            window.confirm(
-              `The develop process needs to be restarted for the changes to ${msg.dirtyFile} to be applied.\nDo you want to restart the develop process now?`
-            )
+            !isRestarting &&
+            msg.type === `LOG_ACTION` &&
+            msg.action.type === `DEVELOP` &&
+            msg.action.payload === `RESTART_REQUIRED`
           ) {
-            parentSocket.once(`develop:is-starting`, msg => {
-              window.location.reload()
-            })
-            parentSocket.once(`develop:started`, msg => {
-              window.location.reload()
-            })
-            parentSocket.emit(`develop:restart`)
+            if (
+              window.confirm(
+                `The develop process needs to be restarted for the changes to ${msg.action.dirtyFile} to be applied.\nDo you want to restart the develop process now?`
+              )
+            ) {
+              isRestarting = true
+              parentSocket.emit(`develop:restart`, () => {
+                window.location.reload()
+              })
```

Comment:
How does the page recover / notify the user if an error happens after `.emit`, causing the callback not to be triggered? 

Maybe a `setTimeout` can be used here as a heartbeat kind of thing? It wouldn't need to be cleared because the `.reload` would do that implicitly. It would allow us to give a cue to the user to recover manually, somehow, rather than beachballing them :beach_umbrella:

Reply:
Great call! Let's do that as part of #26064!